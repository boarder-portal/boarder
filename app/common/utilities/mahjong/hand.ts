import { Tile } from 'common/types/games/mahjong';

import { isEqualTilesCallback } from 'common/utilities/mahjong/tiles';

export function getHandWithoutTile(hand: Tile[], tile: Tile): Tile[] {
  const winningTileIndex = hand.findIndex(isEqualTilesCallback(tile));

  if (winningTileIndex === -1) {
    return hand;
  }

  return [...hand.slice(0, winningTileIndex), ...hand.slice(winningTileIndex + 1)];
}
