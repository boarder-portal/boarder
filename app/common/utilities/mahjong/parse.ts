import {
  FLOWERS_STRING_VALUES,
  STRING_TO_DRAGON_MAP,
  STRING_TO_SUIT_MAP,
  STRING_TO_WIND_MAP,
  SUITED_STRING_VALUES,
} from 'common/constants/games/mahjong';

import { ESet, ESetConcealedType, ETileType, IKongSet, IPungSet, TTile } from 'common/types/mahjong';

export function parseTiles(tilesString: string): TTile[] {
  return tilesString.replace(/\s/g, '').match(/../g)?.map(parseTile) ?? [];
}

export function parseTile(tileString: string): TTile {
  if (tileString.length !== 2) {
    throw new Error(`Wrong length, expected 2, got ${tileString.length}`);
  }

  const [first, second] = tileString;

  if (first === 'f') {
    const value = FLOWERS_STRING_VALUES.indexOf(second);

    if (value === -1) {
      throw new Error(`Wrong flower value, expected 1-8, got ${second}`);
    }

    return {
      type: ETileType.FLOWER,
      index: value,
    };
  }

  if (first === 'D') {
    const dragon = STRING_TO_DRAGON_MAP[second];

    if (!dragon) {
      throw new Error(`Unexpected dragon value, expected one of "r", "g", "w", got "${second}"`);
    }

    return {
      type: ETileType.DRAGON,
      color: dragon,
    };
  }

  if (first === 'W') {
    const wind = STRING_TO_WIND_MAP[second];

    if (!wind) {
      throw new Error(`Unexpected wind value, expected one of "e", "s", "w", "n", got "${second}"`);
    }

    return {
      type: ETileType.WIND,
      side: wind,
    };
  }

  const suit = STRING_TO_SUIT_MAP[first];

  if (!suit) {
    throw new Error(`Unexpected suit value, expected one of "b", "c", "d", got ${first}`);
  }

  const value = SUITED_STRING_VALUES.indexOf(second) + 1;

  if (!value) {
    throw new Error(`Wrong tile value, expected 1-9, got ${second}`);
  }

  return {
    type: ETileType.SUIT,
    suit,
    value,
  };
}

export function parsePung(pungTileString: string, concealedType = ESetConcealedType.CONCEALED): IPungSet {
  return {
    type: ESet.PUNG,
    tiles: parseTiles(pungTileString.repeat(3)),
    concealedType,
  };
}

export function parseKong(kongTileString: string, concealedType = ESetConcealedType.CONCEALED): IKongSet {
  return {
    type: ESet.KONG,
    tiles: parseTiles(kongTileString.repeat(4)),
    concealedType,
  };
}
