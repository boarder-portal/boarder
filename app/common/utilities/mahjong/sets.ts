import sortBy from 'lodash/sortBy';

import { KNITTED_SEQUENCES } from 'common/constants/games/mahjong/tiles';

import {
  ESet,
  ESetConcealedType,
  IChowSet,
  IGameDeclaredConcealedSet,
  IGameDeclaredMeldedSet,
  IKnittedChowSet,
  IKongSet,
  IPairSet,
  IPungSet,
  ISuitedTile,
  TConcealedSet,
  TGameDeclaredSet,
  TMeldedSet,
  TPlayableTile,
  TSet,
  TTile,
} from 'common/types/mahjong';

import {
  getTileSortValue,
  isEqualTiles,
  isEqualTilesCallback,
  isFlush,
  isPlayable,
  isStraight,
  isSuited,
  isTileSubset,
  kong,
  tilesContainTile,
} from 'common/utilities/mahjong/tiles';
import { getCombinations, getSetsCombinations } from 'common/utilities/combinations';

const SET_SORT_VALUES: Record<ESet, number> = {
  [ESet.KNITTED_CHOW]: 0,
  [ESet.KONG]: 1000,
  [ESet.PUNG]: 2000,
  [ESet.CHOW]: 3000,
  [ESet.PAIR]: 4000,
};

export function isPair(set: TSet): set is IPairSet {
  return set.type === ESet.PAIR;
}

export function isPung(set: TSet): set is IPungSet {
  return set.type === ESet.PUNG || set.type === ESet.KONG;
}

export function isKong(set: TSet): set is IKongSet {
  return set.type === ESet.KONG;
}

export function isChow(set: TSet): set is IChowSet {
  return set.type === ESet.CHOW;
}

export function isKnittedChow(set: TSet): set is IKnittedChowSet {
  return set.type === ESet.KNITTED_CHOW;
}

export function isConcealed(set: TSet): set is TConcealedSet {
  return set.concealedType === ESetConcealedType.CONCEALED;
}

export function isMelded(set: TSet): set is TMeldedSet {
  return !isConcealed(set);
}

export function isDeclaredConcealedSet(set: TGameDeclaredSet): set is IGameDeclaredConcealedSet {
  return isConcealed(set.set);
}

export function isDeclaredMeldedSet(set: TGameDeclaredSet): set is IGameDeclaredMeldedSet {
  return isMelded(set.set);
}

export function arePungs(sets: TSet[]): sets is IPungSet[] {
  return sets.every(isPung);
}

export function areKongs(sets: TSet[]): sets is IKongSet[] {
  return sets.every(isKong);
}

export function areChows(sets: TSet[]): sets is IChowSet[] {
  return sets.every(isChow);
}

export function isEqualSets(set1: TSet, set2: TSet): boolean {
  return (
    set1.type === set2.type &&
    set1.tiles.length === set2.tiles.length &&
    set1.tiles.every((tile, index) => isEqualTiles(tile, set2.tiles.at(index)))
  );
}

export function isEqualSetOfSets(sets1: TSet[], sets2: TSet[]): boolean {
  if (sets1.length !== sets2.length) {
    return false;
  }

  const sets2Copy = [...sets2];

  return sets1.every((set) => {
    const equalSetIndex = sets2Copy.findIndex((set2) => isEqualSets(set, set2));

    if (equalSetIndex === -1) {
      return false;
    }

    sets2Copy.splice(equalSetIndex, 1);

    return true;
  });
}

export function getSetTile<Set extends TSet>(set: Set): Set['tiles'][number] {
  return set.type === ESet.CHOW || set.type === ESet.KNITTED_CHOW ? set.tiles[1] : set.tiles[0];
}

export function getSetSortValue(set: TSet): number {
  if (isKnittedChow(set)) {
    return getSetTile(set).value;
  }

  return SET_SORT_VALUES[set.type] + getTileSortValue(getSetTile(set));
}

export interface ISetsVariationsOptions {
  hand: TPlayableTile[];
  declaredSets: TSet[];
  isSelfDraw: boolean;
}

export function getSetsVariations(options: ISetsVariationsOptions): TSet[][] {
  const sortedHand = sortBy(options.hand, getTileSortValue);
  const winningTile = options.hand.at(-1);

  if (!winningTile) {
    return [];
  }

  const setsVariations = splitIntoSets({
    hand: sortedHand,
    pairsFound: 0,
    allPairsAllowed: sortedHand.length === 14,
  });

  return setsVariations
    .flatMap((sets) => {
      if (options.isSelfDraw) {
        return [[...options.declaredSets, ...sets]];
      }

      const variations: TSet[][] = [];

      sets.forEach((set, setIndex) => {
        if (tilesContainTile(set.tiles, winningTile)) {
          variations.push([
            ...sets.slice(0, setIndex),
            {
              ...set,
              concealedType: ESetConcealedType.WINNING_MELDED,
            },
            ...sets.slice(setIndex + 1),
          ]);
        }
      });

      return variations.map((sets) => [...options.declaredSets, ...sets]);
    })
    .map((sets) => sortBy(sets, getSetSortValue));
}

export function getAllSetsCombinations(sets: TSet[]): TSet[][] {
  return getSetsCombinations(sets.map((set) => [[set], []])).map((sets) => sets.flat());
}

interface ISplitSetsOptions {
  hand: TPlayableTile[];
  pairsFound: number;
  allPairsAllowed: boolean;
}

function splitIntoSets(options: ISplitSetsOptions): TSet[][] {
  const { hand, pairsFound, allPairsAllowed } = options;

  const firstTile = hand.at(0);

  if (!firstTile) {
    return [[]];
  }

  const sets: TSet[][] = [];

  if (pairsFound === 0 || allPairsAllowed) {
    const hasPair = isEqualTiles(firstTile, hand.at(1));

    if (hasPair) {
      sets.push(
        ...splitIntoSets({
          hand: hand.slice(2),
          pairsFound: pairsFound + 1,
          allPairsAllowed,
        }).map((sets) => [
          {
            type: ESet.PAIR,
            tiles: hand.slice(0, 2),
            concealedType: ESetConcealedType.CONCEALED,
          } as const,
          ...sets,
        ]),
      );
    }
  }

  if (pairsFound <= 1) {
    const hasPung = isEqualTiles(firstTile, hand.at(2));

    if (hasPung) {
      sets.push(
        ...splitIntoSets({
          hand: hand.slice(3),
          pairsFound,
          allPairsAllowed: false,
        }).map((sets) => [
          {
            type: ESet.PUNG,
            tiles: hand.slice(0, 3),
            concealedType: ESetConcealedType.CONCEALED,
          } as const,
          ...sets,
        ]),
      );
    }

    if (isSuited(firstTile)) {
      const higherBy1TileIndex = hand.findIndex(
        (tile) => isSuited(tile) && tile.suit === firstTile.suit && tile.value === firstTile.value + 1,
      );

      if (higherBy1TileIndex !== -1) {
        const higherBy2TileIndex = hand.findIndex(
          (tile) => isSuited(tile) && tile.suit === firstTile.suit && tile.value === firstTile.value + 2,
        );

        if (higherBy2TileIndex !== -1) {
          sets.push(
            ...splitIntoSets({
              hand: [
                ...hand.slice(1, higherBy1TileIndex),
                ...hand.slice(higherBy1TileIndex + 1, higherBy2TileIndex),
                ...hand.slice(higherBy2TileIndex + 1),
              ],
              pairsFound,
              allPairsAllowed: false,
            }).map((sets) => [
              {
                type: ESet.CHOW,
                tiles: [firstTile, hand[higherBy1TileIndex] as ISuitedTile, hand[higherBy2TileIndex] as ISuitedTile],
                concealedType: ESetConcealedType.CONCEALED,
              },
              ...sets,
            ]),
          );
        }
      }

      if (hand.length >= 9) {
        const knittedSequence = KNITTED_SEQUENCES.find((knittedChows) => {
          return (
            knittedChows.some((knittedChow) => tilesContainTile(knittedChow, firstTile)) &&
            knittedChows.every((knittedChow) => isTileSubset(knittedChow, hand))
          );
        });

        if (knittedSequence) {
          const knittedTiles = knittedSequence.flat();
          const restHand = hand.filter((tile) => {
            const tileIndex = knittedTiles.findIndex(isEqualTilesCallback(tile));

            if (tileIndex === -1) {
              return true;
            }

            knittedTiles.splice(tileIndex, 1);

            return false;
          });

          sets.push(
            ...splitIntoSets({
              hand: restHand,
              pairsFound,
              allPairsAllowed: false,
            }).map((sets) => [
              ...knittedSequence.map((tiles) => ({
                type: ESet.KNITTED_CHOW,
                tiles,
                concealedType: ESetConcealedType.CONCEALED,
              })),
              ...sets,
            ]),
          );
        }
      }
    }
  }

  return sets;
}

export function getTileHogs(sets: TSet[]): TPlayableTile[] {
  const tileHogs: TPlayableTile[] = [];

  sets
    .flatMap(({ tiles }) => tiles)
    .forEach((tile, _, tiles) => {
      if (tilesContainTile(tileHogs, tile)) {
        return;
      }

      if (sets.some((set) => isKong(set) && isEqualTiles(getSetTile(set), tile))) {
        return;
      }

      if (tiles.filter(isEqualTilesCallback(tile)).length === 4) {
        tileHogs.push(tile);
      }
    });

  return tileHogs;
}

export function getPossibleKongs(hand: TTile[], declaredSets: TGameDeclaredSet[]): IKongSet[] {
  const possibleSets: IKongSet[] = [];

  hand.forEach((tile, index) => {
    if (hand.slice(index + 1).filter(isEqualTilesCallback(tile)).length === 3 && isPlayable(tile)) {
      possibleSets.push({
        type: ESet.KONG,
        tiles: kong(tile),
        concealedType: ESetConcealedType.CONCEALED,
      });
    }
  });

  declaredSets.forEach(({ set }) => {
    if (!isPung(set)) {
      return;
    }

    const pungTile = getSetTile(set);

    if (hand.some(isEqualTilesCallback(pungTile))) {
      possibleSets.push({
        type: ESet.KONG,
        tiles: kong(pungTile),
        concealedType: ESetConcealedType.MELDED,
      });
    }
  });

  return possibleSets;
}

export function getPossibleMeldedSets(
  hand: TTile[],
  discardedTile: TPlayableTile,
  isChowPossible: boolean,
): TMeldedSet[] {
  const possibleSets: TMeldedSet[] = [];
  const allPossibleSets = getCombinations(sortBy([...hand, discardedTile], getTileSortValue), 3);
  const isDiscardedTile = isEqualTilesCallback(discardedTile);

  const addSet = (setToAdd: TMeldedSet) => {
    if (possibleSets.every((set) => !isEqualSets(set, setToAdd))) {
      possibleSets.push(setToAdd);
    }
  };

  allPossibleSets.forEach((tiles) => {
    if (tiles.every(isDiscardedTile) && tiles.every(isPlayable)) {
      addSet({
        type: ESet.PUNG,
        tiles,
        concealedType: ESetConcealedType.MELDED,
      });
    } else if (isChowPossible && tiles.some(isDiscardedTile) && isFlush(tiles) && isStraight(tiles)) {
      addSet({
        type: ESet.CHOW,
        tiles,
        concealedType: ESetConcealedType.MELDED,
      });
    }
  });

  if (hand.filter(isDiscardedTile).length === 3) {
    possibleSets.push({
      type: ESet.KONG,
      tiles: kong(discardedTile),
      concealedType: ESetConcealedType.MELDED,
    });
  }

  return sortBy(possibleSets, getSetSortValue);
}
