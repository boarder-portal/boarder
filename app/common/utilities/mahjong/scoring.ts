import sortBy from 'lodash/sortBy';

import { STANDARD_TILES } from '../../constants/games/mahjong';
import { FAN_SCORES, NO_SETS_FANS } from 'common/constants/games/mahjong/fans';

import { EFan, EFanType, EWind, TSet, TTile } from 'common/types/mahjong';

import { getSetsCombinations, getSetsVariations } from 'common/utilities/mahjong/sets';
import {
  canAddFan,
  getFansScore,
  getSetsFans,
  getSpecialFans,
  getSpecialSetsFans,
  getWholeHandFans,
  getWholeHandSetsFans,
} from 'common/utilities/mahjong/fans';

export interface IHandScoreOptions {
  hand: TTile[];
  concealedSets: TSet[];
  meldedSets: TSet[];
  flowers: TTile[];
  seatWind: EWind;
  roundWind: EWind;
  isSelfDraw: boolean;
  isReplacementTile: boolean;
  isRobbingKong: boolean;
  isLastTile: boolean;
  isLastWallTile: boolean;
}

export interface IHandScoreFullOptions extends IHandScoreOptions {
  winningTile: TTile;
  waits: TTile[];
}

export interface IHandFan {
  type: EFanType.HAND;
  fan: EFan;
}

export interface ISetsFan {
  type: EFanType.SETS;
  fan: EFan;
  sets: TSet[];
}

export interface ISpecialFan {
  type: EFanType.SPECIAL;
  fan: EFan;
  tiles: TTile[];
}

export type TFan = IHandFan | ISetsFan | ISpecialFan;

export interface IHandMahjong {
  fans: TFan[];
  score: number;
}

export function getAllWaits(options: IHandScoreOptions): TTile[] {
  const waitsWithSingleWait = STANDARD_TILES.filter((tile) =>
    getHandMahjong({
      ...options,
      winningTile: tile,
      waits: [tile],
    }),
  );
  const waitsWithoutSingleWait = STANDARD_TILES.filter((tile) =>
    getHandMahjong({
      ...options,
      winningTile: tile,
      waits: STANDARD_TILES,
    }),
  );

  if (!waitsWithoutSingleWait.length) {
    return waitsWithSingleWait.length === 1 ? waitsWithSingleWait : [];
  }

  return waitsWithoutSingleWait;
}

export function getHandMahjong(options: IHandScoreFullOptions): IHandMahjong | null {
  const { hand, concealedSets, meldedSets, winningTile, seatWind, roundWind } = options;
  const knownSets = [...concealedSets, ...meldedSets];

  if (knownSets.length * 3 + hand.length !== 13) {
    return null;
  }

  const wholeHand = [
    ...hand,
    ...concealedSets.flatMap(({ tiles }) => tiles),
    ...meldedSets.flatMap(({ tiles }) => tiles),
    winningTile,
  ];
  const wholeHandFans = getWholeHandFans(wholeHand);
  const specialFans = getSpecialFans(options);

  const setsVariations = getSetsVariations({
    hand: [...hand, winningTile],
    knownSets: [...concealedSets, ...meldedSets],
    isSelfDraw: options.isSelfDraw,
  });

  let mahjong: IHandMahjong | null = null;

  if (
    setsVariations.length === 0 &&
    NO_SETS_FANS.every((fanType) => wholeHandFans.every(({ fan }) => fan !== fanType))
  ) {
    return null;
  }

  console.log(setsVariations);

  if (setsVariations.length > 0) {
    setsVariations.forEach((sets) => {
      const wholeHandSetsFans = getWholeHandSetsFans(sets, options.isSelfDraw);
      const specialSetsFans = getSpecialSetsFans(sets, winningTile, options.waits);
      const setsFans: TFan[] = [];

      getSetsCombinations(sets).forEach((sets) => {
        setsFans.push(...getSetsFans(sets, seatWind, roundWind));
      });

      const fans = [...wholeHandFans, ...wholeHandSetsFans, ...setsFans, ...specialFans, ...specialSetsFans];

      if (fans.every((fan) => fan.fan === EFan.FLOWER_TILES)) {
        fans.push({
          type: EFanType.HAND,
          fan: EFan.CHICKEN_HAND,
        });
      }

      const fansMahjong = getBestFansMahjong(fans);

      if (!mahjong || (fansMahjong && fansMahjong.score > mahjong.score)) {
        mahjong = fansMahjong;
      }
    });
  } else {
    mahjong = getBestFansMahjong([...wholeHandFans, ...specialFans]);
  }

  if (!mahjong) {
    return null;
  }

  return mahjong;
}

function getBestFansMahjong(fans: TFan[]): IHandMahjong | null {
  fans = sortBy(fans, ({ fan }) => -FAN_SCORES[fan]);

  let pickedFans = null as TFan[] | null;
  let maxScore = 0;

  const calculateBestMahjong = (fansLeft: TFan[], includedFans: TFan[]): void => {
    const firstFan = fansLeft.at(0);

    if (!firstFan) {
      const score = getFansScore(includedFans);

      if (score > maxScore || !pickedFans) {
        maxScore = score;
        pickedFans = includedFans;
      }

      return;
    }

    const restFans = fansLeft.slice(1);

    if (canAddFan(includedFans, firstFan)) {
      calculateBestMahjong(restFans, [...includedFans, firstFan]);
    }

    const maxPossibleScore = getFansScore([...includedFans, ...restFans]);

    if (maxPossibleScore > maxScore) {
      calculateBestMahjong(restFans, includedFans);
    }
  };

  calculateBestMahjong(fans, []);

  if (!pickedFans) {
    return null;
  }

  const pureFans = pickedFans.filter(({ fan }) => fan !== EFan.FLOWER_TILES);

  if (getFansScore(pureFans) < 8) {
    return null;
  }

  return { fans: pickedFans, score: getFansScore(pickedFans) };
}
