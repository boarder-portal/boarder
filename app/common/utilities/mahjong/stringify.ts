import {
  DRAGON_TO_STRING_MAP,
  FLOWERS_STRING_VALUES,
  SET_NAMES as HUMAN_SET_NAMES,
  SUITED_STRING_VALUES,
  SUIT_TO_STRING_MAP,
  WIND_NAMES,
  WIND_SHORT_NAMES,
  WIND_TO_STRING_MAP,
} from 'common/constants/games/mahjong';

import { Fan, HandMahjong, Set, SetConcealedType, SetType, Tile, TileType, WindSide } from 'common/types/mahjong';

import { isSetsFan, isSpecialFan } from 'common/utilities/mahjong/fans';
import { getSetTile } from 'common/utilities/mahjong/sets';

const SET_NAMES: Record<SetType, string> = {
  [SetType.PAIR]: 'pair',
  [SetType.PUNG]: 'pung',
  [SetType.KONG]: 'kong',
  [SetType.CHOW]: 'chow',
  [SetType.KNITTED_CHOW]: 'knit',
};

export function stringifyTile(tile: Tile): string {
  if (tile.type === TileType.FLOWER) {
    return `f${FLOWERS_STRING_VALUES[tile.index]}`;
  }

  if (tile.type === TileType.DRAGON) {
    return `D${DRAGON_TO_STRING_MAP[tile.color]}`;
  }

  if (tile.type === TileType.WIND) {
    return `W${WIND_TO_STRING_MAP[tile.side]}`;
  }

  return `${SUIT_TO_STRING_MAP[tile.suit]}${SUITED_STRING_VALUES[tile.value - 1]}`;
}

export function stringifySet(set: Set): string {
  return `${set.concealedType === SetConcealedType.CONCEALED ? 'c' : 'm'}_${SET_NAMES[set.type]} ${stringifyTile(
    getSetTile(set),
  )}`;
}

export function stringifyFan(fan: Fan): string {
  if (isSetsFan(fan)) {
    return `${fan.fan} (${fan.sets.map(stringifySet).join(', ')})`;
  }

  if (isSpecialFan(fan)) {
    return `${fan.fan}${fan.tile ? ` (${stringifyTile(fan.tile)})` : ''}`;
  }

  return fan.fan;
}

export function stringifyMahjong(mahjong: HandMahjong | null): string {
  if (!mahjong) {
    return 'null';
  }

  return `${mahjong.score}

${mahjong.waits.map(stringifyTile).join(' ')}

${mahjong.fans.map(stringifyFan).join('\n')}${mahjong.sets ? `\n\n${mahjong.sets.map(stringifySet).join('\n')}` : ''}`;
}

export function getSetNumanName(set: Set): string {
  return HUMAN_SET_NAMES[set.type];
}

export function getWindHumanName(wind: WindSide): string {
  return WIND_NAMES[wind];
}

export function getWindHumanShortName(wind: WindSide): string {
  return WIND_SHORT_NAMES[wind];
}
