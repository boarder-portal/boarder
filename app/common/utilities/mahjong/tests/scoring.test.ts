import { ESetConcealedType, EWind } from 'common/types/mahjong';

import { getHandMahjong } from 'common/utilities/mahjong/scoring';
import { parseChow, parseKong, parsePung, parseTile, parseTiles } from 'common/utilities/mahjong/parse';
import { stringifyMahjong } from 'common/utilities/mahjong/stringify';

const standardOptions = {
  roundWind: EWind.EAST,
  seatWind: EWind.EAST,
  isSelfDraw: true,
  meldedSets: [],
  concealedSets: [],
  flowers: [],
  isLastTile: false,
  isRobbingKong: false,
  isReplacementTile: false,
  isLastWallTile: false,
};

describe('mahjong', () => {
  describe('scoring', () => {
    describe('basic scoring', () => {
      test('basic mahjong', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              hand: parseTiles('WeWeWe WsWsWs WwWwWw WnWnWn Dr'),
              waits: parseTiles('Dr'),
              winningTile: parseTile('Dr'),
            }),
          ),
        ).toMatchSnapshot();
      });
    });

    describe('big four winds', () => {
      test('half flush', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              meldedSets: [parsePung('We', ESetConcealedType.MELDED), parsePung('Ws', ESetConcealedType.MELDED)],
              hand: parseTiles('WwWwWw WnWn c6c6'),
              waits: parseTiles('Wnc6'),
              winningTile: parseTile('Wn'),
            }),
          ),
        ).toMatchSnapshot();
      });

      test('terminals', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              hand: parseTiles('WeWeWe WsWsWs WwWwWw WnWn b9b9'),
              waits: parseTiles('Wnb9'),
              winningTile: parseTile('Wn'),
              isSelfDraw: false,
            }),
          ),
        ).toMatchSnapshot();
      });

      test('honors', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              concealedSets: [parseKong('We', ESetConcealedType.CONCEALED)],
              meldedSets: [parsePung('Ws', ESetConcealedType.MELDED)],
              hand: parseTiles('WwWwWw WnWn DrDr'),
              waits: parseTiles('WnDr'),
              winningTile: parseTile('Wn'),
            }),
          ),
        ).toMatchSnapshot();
      });
    });

    describe('big three dragons', () => {
      test('half flush', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              meldedSets: [
                parsePung('Dr', ESetConcealedType.MELDED),
                parsePung('Dg', ESetConcealedType.MELDED),
                parsePung('Dw', ESetConcealedType.MELDED),
              ],
              hand: parseTiles('c5c5c6c7'),
              waits: parseTiles('c5c8'),
              winningTile: parseTile('c5'),
            }),
          ),
        ).toMatchSnapshot();
      });

      test('terminals', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              meldedSets: [
                parsePung('Dr', ESetConcealedType.MELDED),
                parsePung('Dg', ESetConcealedType.MELDED),
                parsePung('Dw', ESetConcealedType.MELDED),
                parsePung('d9', ESetConcealedType.MELDED),
              ],
              hand: parseTiles('b1'),
              waits: parseTiles('b1'),
              winningTile: parseTile('b1'),
              isSelfDraw: false,
            }),
          ),
        ).toMatchSnapshot();
      });

      test('honors', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              hand: parseTiles('DrDrDr DgDgDg DwDwDw WwWwWw Wn'),
              waits: parseTiles('Wn'),
              winningTile: parseTile('Wn'),
              isSelfDraw: false,
              roundWind: EWind.WEST,
              seatWind: EWind.WEST,
            }),
          ),
        ).toMatchSnapshot();
      });
    });

    describe('all green', () => {
      test('pure shifted pungs', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              meldedSets: [
                parsePung('b2', ESetConcealedType.MELDED),
                parsePung('b3', ESetConcealedType.MELDED),
                parsePung('b4', ESetConcealedType.MELDED),
              ],
              hand: parseTiles('DgDgb6b6'),
              waits: parseTiles('Dgb6'),
              winningTile: parseTile('Dg'),
            }),
          ),
        ).toMatchSnapshot();
      });

      test('triple chow', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              meldedSets: [
                parseChow('b3', ESetConcealedType.MELDED),
                parseChow('b3', ESetConcealedType.MELDED),
                parseChow('b3', ESetConcealedType.MELDED),
                parsePung('b8', ESetConcealedType.MELDED),
              ],
              hand: parseTiles('b6'),
              waits: parseTiles('b6'),
              winningTile: parseTile('b6'),
              isSelfDraw: false,
            }),
          ),
        ).toMatchSnapshot();
      });

      test('seven pairs', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              hand: parseTiles('b2b2b3b3b4b4b6b6b8b8b8b8Dg'),
              waits: parseTiles('Dg'),
              winningTile: parseTile('Dg'),
            }),
          ),
        ).toMatchSnapshot();
      });
    });

    describe('nine gates', () => {
      test('edge', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              hand: parseTiles('c1c1c1c2c3c4c5c6c7c8c9c9c9'),
              waits: parseTiles('c1c2c3c4c5c6c7c8c9'),
              winningTile: parseTile('c9'),
            }),
          ),
        ).toMatchSnapshot();
      });

      test('middle', () => {
        expect(
          stringifyMahjong(
            getHandMahjong({
              ...standardOptions,
              hand: parseTiles('c1c1c1c2c3c4c5c6c7c8c9c9c9'),
              waits: parseTiles('c1c2c3c4c5c6c7c8c9'),
              winningTile: parseTile('c5'),
            }),
          ),
        ).toMatchSnapshot();
      });
    });
  });
});
