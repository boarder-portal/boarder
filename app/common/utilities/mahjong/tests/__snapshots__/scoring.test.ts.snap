// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mahjong scoring all green pure shifted pungs 1`] = `
"127

ALL_GREEN
PURE_SHIFTED_PUNGS (m_pung b2, m_pung b3, m_pung b4)
HALF_FLUSH
ALL_PUNGS
DRAGON_PUNG (c_pung Dg)
SELF_DRAWN

m_pung b2
m_pung b3
m_pung b4
c_pung Dg
c_pair b6"
`;

exports[`mahjong scoring all green seven pairs 1`] = `
"124

ALL_GREEN
SEVEN_PAIRS
HALF_FLUSH
FULLY_CONCEALED_HAND
TILE_HOG (b8)

c_pair b2
c_pair b3
c_pair b4
c_pair b6
c_pair b8
c_pair b8
c_pair Dg"
`;

exports[`mahjong scoring all green triple chow 1`] = `
"144

ALL_GREEN
FULL_FLUSH
PURE_TRIPLE_CHOW (m_chow b3, m_chow b3, m_chow b3)
MELDED_HAND
ALL_SIMPLES

m_chow b3
m_chow b3
m_chow b3
m_pung b8
m_pair b6"
`;

exports[`mahjong scoring basic scoring basic mahjong 1`] = `
"221

BIG_FOUR_WINDS (c_pung We, c_pung Ws, c_pung Ww, c_pung Wn)
ALL_HONORS
FOUR_CONCEALED_PUNGS (c_pung We, c_pung Ws, c_pung Ww, c_pung Wn)
FULLY_CONCEALED_HAND
SINGLE_WAIT

c_pung We
c_pung Ws
c_pung Ww
c_pung Wn
c_pair Dr"
`;

exports[`mahjong scoring big four winds half flush 1`] = `
"97

BIG_FOUR_WINDS (m_pung We, m_pung Ws, c_pung Ww, c_pung Wn)
HALF_FLUSH
TWO_CONCEALED_PUNGS (c_pung Ww, c_pung Wn)
SELF_DRAWN

m_pung We
m_pung Ws
c_pung Ww
c_pung Wn
c_pair c6"
`;

exports[`mahjong scoring big four winds honors 1`] = `
"171

BIG_FOUR_WINDS (c_kong We, m_pung Ws, c_pung Ww, c_pung Wn)
ALL_HONORS
THREE_CONCEALED_PUNGS (c_kong We, c_pung Ww, c_pung Wn)
CONCEALED_KONG (c_kong We)
SELF_DRAWN

c_kong We
m_pung Ws
c_pung Ww
c_pung Wn
c_pair Dr"
`;

exports[`mahjong scoring big four winds terminals 1`] = `
"144

BIG_FOUR_WINDS (c_pung We, c_pung Ws, c_pung Ww, m_pung Wn)
ALL_TERMINALS_AND_HONORS
THREE_CONCEALED_PUNGS (c_pung We, c_pung Ws, c_pung Ww)
HALF_FLUSH
CONCEALED_HAND

c_pung We
c_pung Ws
c_pung Ww
m_pung Wn
c_pair b9"
`;

exports[`mahjong scoring big three dragons half flush 1`] = `
"95

BIG_THREE_DRAGONS (m_pung Dr, m_pung Dg, m_pung Dw)
HALF_FLUSH
SELF_DRAWN

m_pung Dr
m_pung Dg
m_pung Dw
c_chow c6
c_pair c5"
`;

exports[`mahjong scoring big three dragons honors 1`] = `
"221

BIG_THREE_DRAGONS (c_pung Dg, c_pung Dr, c_pung Dw)
ALL_HONORS
FOUR_CONCEALED_PUNGS (c_pung Ww, c_pung Dg, c_pung Dr, c_pung Dw)
PREVALENT_WIND (c_pung Ww)
SEAT_WIND (c_pung Ww)
SINGLE_WAIT

c_pung Ww
c_pung Dg
c_pung Dr
c_pung Dw
m_pair Wn"
`;

exports[`mahjong scoring big three dragons terminals 1`] = `
"127

BIG_THREE_DRAGONS (m_pung Dr, m_pung Dg, m_pung Dw)
ALL_TERMINALS_AND_HONORS
MELDED_HAND
ONE_VOIDED_SUIT

m_pung Dr
m_pung Dg
m_pung Dw
m_pung d9
m_pair b1"
`;

exports[`mahjong scoring nine gates edge 1`] = `
"110

NINE_GATES
PURE_STRAIGHT (c_chow c2, c_chow c5, c_chow c8)
FULLY_CONCEALED_HAND
TILE_HOG (c9)

c_pung c9
c_chow c2
c_chow c5
c_chow c8
c_pair c1"
`;

exports[`mahjong scoring nine gates middle 1`] = `
"94

NINE_GATES
FULLY_CONCEALED_HAND
TWO_CONCEALED_PUNGS (c_pung c1, c_pung c9)

c_pung c1
c_pung c9
c_chow c3
c_chow c7
c_pair c5"
`;
